testbed/interfaces/node.go

```diff
@@ -0,0 +1,145 @@
+package testbedi
+
+import (
+	"context"
+	"io"
+)
+
+// NewNodeFunc constructs a node implementing the Core interface. It is provided
+// a path to an already created directory `dir`, as well as a map of attributes
+// which can be supplied to shape process execution.
+// Examples of attributes include: which binary to use, docker image, cpu/ram
+// limits, or any other information that may be required to property setup or
+// manage the node.
+type NewNodeFunc func(dir string, attrs map[string]interface{}) (Core, error)
+
+// GetAttrListFunc returns a list of attribute names that can be queried from
+// the node. These attributes may include those can be set from the NewNodeFunc,
+// or additional attributes at may only be available after initialization.
+// Attributes returned should be queriable through the Attribute interface.
+// Examples include: api address, peerid, cpu/ram limits, jitter.
+type GetAttrListFunc func() []string
+
+// GetAttrDescFunc returns the description of the attribute `attr`
+type GetAttrDescFunc func(attr string) (string, error)
+
+type Libp2p interface {
+	// PeerID returns the peer id
+	PeerID() (string, error)
+	// APIAddr returns the multiaddr for the api
+	APIAddr() (string, error)
+	// SwarmAddrs returns the swarm addrs for the node
+	SwarmAddrs() ([]string, error)
```

Comment:
Should probably use multiaddrs.

Reply:
Going to answer these three here. Originally we had both cids and multiaddrs on the interface. Ended up stripping both of them primarily due to an issue around multiaddrs (though maybe it's not really an issue).

Basically, because the primary use of the interface is for plugins. If a plugin uses a multiaddr codec that the loader doesn't have, panics are bound to occur.

https://github.com/multiformats/go-multiaddr/blob/master/multiaddr.go#L64
https://github.com/multiformats/go-multiaddr/blob/master/multiaddr.go#L84

This isn't so much an issue with iptb, as it doesn't directly use this interface at the moment (used primarily inside of `connect`).

Maybe the panics are fine, and it up to the consumer to recover from them, I'm open to moving back as long as we thing it's better.